### PR TITLE
local: fixes a data race in anti-entropy sync

### PR DESCRIPTION
**Pull request by rboyer**
The race detector noticed this initially in `TestAgentConfigWatcherSidecarProxy` but it is not restricted to just tests.

The two main changes here were:

- ensure that before we mutate the internal `agent/local` representation of a Service (for tags or VIPs) we clone those fields
- ensure that there's no function argument joint ownership between the caller of a function and the local state when calling `AddService`, `AddCheck`, and related using `copystructure` for now.

```
Failed
=== RUN   TestAgentConfigWatcherSidecarProxy
[INFO] freeport: detected ephemeral port range of [32768, 60999]
[INFO] freeport: reducing max blocks from 30 to 15 to avoid the ephemeral port range
==================
WARNING: DATA RACE
Write at 0x00c000d7e958 by goroutine 130:
  github.com/hashicorp/consul/agent/local.(*State).updateSyncState()
      /home/circleci/project/agent/local/state.go:1109 +0x109e
  github.com/hashicorp/consul/agent/local.(*State).SyncFull()
      /home/circleci/project/agent/local/state.go:1191 +0x30
  github.com/hashicorp/consul/agent/ae.(*StateSyncer).nextFSMState()
      /home/circleci/project/agent/ae/ae.go:178 +0xa2
  github.com/hashicorp/consul/agent/ae.(*StateSyncer).nextFSMState-fm()
      /home/circleci/project/agent/ae/ae.go:171 +0x4d
  github.com/hashicorp/consul/agent/ae.(*StateSyncer).runFSM()
      /home/circleci/project/agent/ae/ae.go:164 +0x89
  github.com/hashicorp/consul/agent/ae.(*StateSyncer).Run()
      /home/circleci/project/agent/ae/ae.go:158 +0x48
  github.com/hashicorp/consul/agent.(*Agent).StartSync·dwrap·28()
      /home/circleci/project/agent/agent.go:1616 +0x39

Previous read at 0x00c000d7e958 by goroutine 167:
  github.com/hashicorp/consul/agent.buildAgentService()
      /home/circleci/project/agent/agent_endpoint.go:260 +0x1a4
  github.com/hashicorp/consul/agent.(*HTTPHandlers).AgentService.func1()
      /home/circleci/project/agent/agent_endpoint.go:448 +0x29d
  github.com/hashicorp/consul/agent.(*Agent).LocalBlockingQuery()
      /home/circleci/project/agent/agent.go:3842 +0x2d0
  github.com/hashicorp/consul/agent.(*HTTPHandlers).AgentService()
      /home/circleci/project/agent/agent_endpoint.go:423 +0x411
  github.com/hashicorp/consul/agent.(*HTTPHandlers).handler.func3()
      /home/circleci/project/agent/http.go:292 +0x65
  github.com/hashicorp/consul/agent.(*HTTPHandlers).wrap.func1()
      /home/circleci/project/agent/http.go:548 +0x1029
  github.com/hashicorp/consul/agent.(*HTTPHandlers).handler.func1.1()
      /home/circleci/project/agent/http.go:226 +0xcb
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2047 +0x4d
  github.com/NYTimes/gziphandler.GzipHandlerWithOpts.func1.1()
      /home/circleci/go/pkg/mod/github.com/!n!y!times/gziphandler@v1.0.1/gzip.go:287 +0x433
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2047 +0x4d
  net/http.(*ServeMux).ServeHTTP()
      /usr/local/go/src/net/http/server.go:2425 +0xc5
  github.com/hashicorp/go-cleanhttp.PrintablePathCheckHandler.func1()
      /home/circleci/go/pkg/mod/github.com/hashicorp/go-cleanhttp@v0.5.1/handlers.go:42 +0xc1
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2047 +0x4d
  github.com/hashicorp/consul/agent.(*wrappedMux).ServeHTTP()
      /home/circleci/project/agent/http.go:155 +0x69
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2879 +0x89a
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1930 +0x12e4
  net/http.(*Server).Serve·dwrap·87()
      /usr/local/go/src/net/http/server.go:3034 +0x58

Goroutine 130 (running) created at:
  github.com/hashicorp/consul/agent.(*Agent).StartSync()
      /home/circleci/project/agent/agent.go:1616 +0xb8
  github.com/hashicorp/consul/agent.(*TestAgent).Start()
      /home/circleci/project/agent/testagent.go:230 +0xcaa
  github.com/hashicorp/consul/agent.StartTestAgent.func1()
      /home/circleci/project/agent/testagent.go:106 +0x53
  github.com/hashicorp/consul/sdk/testutil/retry.run.func2()
      /home/circleci/project/sdk/testutil/retry/retry.go:148 +0x58
  github.com/hashicorp/consul/sdk/testutil/retry.run()
      /home/circleci/project/sdk/testutil/retry/retry.go:149 +0xfe
  github.com/hashicorp/consul/sdk/testutil/retry.RunWith()
      /home/circleci/project/sdk/testutil/retry/retry.go:108 +0x72
  github.com/hashicorp/consul/agent.StartTestAgent()
      /home/circleci/project/agent/testagent.go:104 +0x13b
  github.com/hashicorp/consul/connect/proxy.TestAgentConfigWatcherSidecarProxy()
      /home/circleci/project/connect/proxy/config_test.go:90 +0x164
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 167 (running) created at:
  net/http.(*Server).Serve()
      /usr/local/go/src/net/http/server.go:3034 +0x847
  github.com/hashicorp/consul/agent.newAPIServerHTTP.func1()
      /home/circleci/project/agent/apiserver.go:104 +0x78
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /home/circleci/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x96
```

### CODE REVIEW

**Comment by rboyer** [comment]
Related: https://github.com/hashicorp/consul/pull/12028 & https://github.com/hashicorp/consul/pull/12041

File: agent/local/state.go [hunk 1]
```diff
@@ -1106,6 +1115,13 @@ func (l *State) updateSyncState() error {
 					m[k] = v
 				}
 			}
+			if !wasCopied {
+				// Make a shallow copy since we're going to mutate it and other
+				// readers may be reading it and we want to avoid a race.
+				wasCopied = true
```
Comment (dnephin) [inline agent/local/state.go:1121]: Is this value ever read after this point? I'm not seeing a loop around this, but maybe I'm missing it?

Can we remove this line?

```suggestion
```

File: .changelog/12324.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+```release-note:bug
+fix data race in TestAgentConfigWatcherSidecarProxy
```
Comment (dnephin) [inline .changelog/12324.txt:2]: Do we usually include changelog entries for test fixes? I thought these were only for things that had user impact.

**Review by dnephin — COMMENTED** [review]
Thanks for working on this! I think @kisunji has a lot more context here, so I'll wait on his review.

Does this depend on both previous fixes, or can we roll back some of that?

File: .changelog/12324.txt (hunk 2 above)
Comment (rboyer) [reply .changelog/12324.txt:2]: There's an actual production race in the data that the test just happens to surface so i figured it was worth calling out. This is not a race in a test per se.

Comment (dnephin) [reply .changelog/12324.txt:2]: Ah, I did not realize that from the changelog. Maybe something like:

> local: fixes a data race in anti-entropy sync that could cause the wrong tags to be applied to a service when EnableTagOverride is used

No idea if that is accurate, but something along those lines would probably help future readers understand the bug

Comment (rboyer) [reply .changelog/12324.txt]: updated

**Comment by rboyer** [comment]
> Does this depend on both previous fixes, or can we roll back some of that?

Parts of the prior fixes are still relevant so I think it makes more sense to roll forward.

File: agent/local/state.go [hunk 3]
```diff
@@ -1083,31 +1083,35 @@ func (l *State) updateSyncState() error {
 			continue
 		}
 
+		// Make a shallow copy since we may mutate it below and other readers
+		// may be reading it and we want to avoid a race.
+		nextService := *ls.Service
```
Comment (rboyer) [inline agent/local/state.go:1088]: This still copies even if it's not going to mutate it for 1 of the 2 reasons below, but I sense that the clarity this provides in reading might be fine given the slightly increased allocations in the AE loop.

File: agent/local/state.go [hunk 4]
```diff
@@ -1083,31 +1083,35 @@ func (l *State) updateSyncState() error {
...
 			for k, v := range rs.TaggedAddresses {
 				if strings.HasPrefix(k, structs.MetaKeyReservedPrefix) {
 					m[k] = v
 				}
 			}
-			ls.Service.TaggedAddresses = m
+			nextService.TaggedAddresses = m
 		}
+
+		ls.Service = &nextService
```
Comment (rboyer) [inline agent/local/state.go:1114]: The `ls` struct is shallow-copied elsewhere when returned to the reading-side of the data race, so the `ls.Service` field should not be subject to a data race.

Access to the map that stores these things is gated by a read lock that this loop holds, so i suspect this particular fix should resolve the rest of the race.

**Comment by kisunji** [comment]
closes https://github.com/hashicorp/consul/issues/12274

**Review by kisunji — APPROVED** [review]

**Review by kisunji — APPROVED** [review]

**Comment by hc-github-team-consul-core** [comment]
:cherries::x: Cherry pick of commit fa4577d1a97d6a165e9d794e7edd5c779a01bf06 onto `release/1.11.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/582260)

**Pull request merged**